In [1]:
# Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Import data
df = pd.read_excel('../data/Online Retail.xlsx')

# 1. Data Preparation

### 1.1 Show dataset

In [3]:
df.head(15)

InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5     536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6     536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7     536366     22633               HAND WARMER UNION JACK         6   
8     536366     22632            HAND WARMER RED POLKA DOT         6   
9     536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   
10    536367     22745           POPPY'S PLAYHOUSE BEDROOM          6   
11    536367     22748            POPPY'S PLAYHOUSE KITCHEN         6   
12    536367     22749    FELTCRAFT PRINCESS CHARLOTTE DOLL         8   
13    536367     22310              IVORY KNITTED MUG COSY          6   
14    536367     84969   BOX OF 6 ASSORTED COLOUR TEASPOONS         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2  2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
5  2010-12-01 08:26:00       7.65     17850.0  United Kingdom  
6  2010-12-01 08:26:00       4.25     17850.0  United Kingdom  
7  2010-12-01 08:28:00       1.85     17850.0  United Kingdom  
8  2010-12-01 08:28:00       1.85     17850.0  United Kingdom  
9  2010-12-01 08:34:00       1.69     13047.0  United Kingdom  
10 2010-12-01 08:34:00       2.10     13047.0  United Kingdom  
11 2010-12-01 08:34:00       2.10     13047.0  United Kingdom  
12 2010-12-01 08:34:00       3.75     13047.0  United Kingdom  
13 2010-12-01 08:34:00       1.65     13047.0  United Kingdom  
14 2010-12-01 08:34:00       4.25     13047.0  United Kingdom

In [4]:
df.shape

(541909, 8)

There are 541,909 rows and 8 columns in the original dataset

### 1.2 Check missing values

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [6]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In the dataset, we encounter two columns - "Description" and "CustomerID" - that contain missing values. Given that we cannot reasonably infer missing "CustomerID" values - perhaps these visitors have not registered or logged into the website - we've decided to remove these rows. Our analysis will then concentrate on customers who have logged in and are identifiable via the website.

### 1.3 Delete rows without customer_id

In [7]:
df.dropna(subset=['CustomerID'], inplace = True)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 406829 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    406829 non-null  object        
 1   StockCode    406829 non-null  object        
 2   Description  406829 non-null  object        
 3   Quantity     406829 non-null  int64         
 4   InvoiceDate  406829 non-null  datetime64[ns]
 5   UnitPrice    406829 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      406829 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 27.9+ MB


In [9]:
df.isna().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

### 1.4 Check and remove duplicated observations

In [10]:
df.duplicated().sum()

5225

In [11]:
df.drop_duplicates(inplace = True)

In [12]:
df.duplicated().sum()

0

In [13]:
df.shape

(401604, 8)

The cleaning dataset contains 401604 rows and 8 columns without any missing values

# 2. Data Type Conversion

In [43]:
df_cleaning = df.copy()

In [44]:
df_cleaning.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

### 2.1 InvoiceNo: Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation. 

To handle the InvoiceNo column that contains both integers and strings, it's necessary to first convert the entire column to a uniform data type, specifically, to string.

In [45]:
# Convert the column to string
df_cleaning['InvoiceNo'] = df_cleaning['InvoiceNo'].astype('str')

df_cleaning['InvoiceNo'].dtypes

dtype('O')

In [46]:
# Check cancealled orders
df_cancelled = df_cleaning[df_cleaning['InvoiceNo'].str.contains('C')]

df_cancelled

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

               InvoiceDate  UnitPrice  CustomerID         Country  
141    2010-12-01 09:41:00      27.50     14527.0  United Kingdom  
154    2010-12-01 09:49:00       4.65     15311.0  United Kingdom  
235    2010-12-01 10:24:00       1.65     17548.0  United Kingdom  
236    2010-12-01 10:24:00       0.29     17548.0  United Kingdom  
237    2010-12-01 10:24:00       0.29     17548.0  United Kingdom  
...                    ...        ...         ...             ...  
540449 2011-12-09 09:57:00       0.83     14397.0  United Kingdom  
541541 2011-12-09 10:28:00     224.69     15498.0  United Kingdom  
541715 2011-12-09 11:57:00      10.95     15311.0  United Kingdom  
541716 2011-12-09 11:58:00       1.25     17315.0  United Kingdom  
541717 2011-12-09 11:58:00       1.25     17315.0  United Kingdom  

[8872 rows x 8 columns]

In relation to the attribute 'Invoice Number', it's important to note that any invoice numbers that begin with a 'C' are used to signify a cancellation.

### 2.2 StockCode: Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.

To handle the StockCode column that contains both integers and strings, it's necessary to first convert the entire column to a uniform data type, specifically, to string.

In [47]:
df_cleaning['StockCode'] = df_cleaning['StockCode'].astype('str')

df_cleaning['StockCode'].dtypes

dtype('O')

### 2.3 Description: Product (item) name. Nominal.

To handle the Description column that contains both integers and strings, it's necessary to first convert the entire column to a uniform data type, specifically, to string.

In [48]:
df_cleaning['Description'] = df_cleaning['Description'].astype('str')

df_cleaning['Description'].dtypes

dtype('O')

### 2.4 Quantity: The quantities of each product (item) per transaction. Numeric.

There's no necessity to alter the data type of this particular column.

### 2.5 InvoiceDate: Invoice Date and time. Numeric, the day and time when each transaction was generated.

There's no necessity to alter the data type of this particular column.

### 2.6 UnitPrice: Unit price. Numeric, Product price per unit in sterling.

There's no necessity to alter the data type of this particular column.

### 2.7 CustomerID: Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.

Change the column of CustomerID from float to integer, and then to string

In [49]:
df_cleaning['CustomerID'] = df_cleaning['CustomerID'].astype('int').astype('str')

df_cleaning['CustomerID'].dtypes

dtype('O')

In [50]:
df_cleaning[['CustomerID']].head()

CustomerID
0      17850
1      17850
2      17850
3      17850
4      17850

### 2.8 Country: Country name. Nominal, the name of the country where each customer resides. 

There's no necessity to alter the data type of this particular column.

# 3. Observations Deletion

1. The presence of negative quantities indicates that the corresponding items were purchased prior to the given transaction date. However, our dataset only contains records from December 2010 onwards. If there's no match for these records, it's likely that the original purchase occurred before December 2010. Therefore, such records should be considered inaccurate for our analysis and we should remove them. 
2. Alternatively, one cancelled order may correspond to one or more previous orders, or orders might exist before or after December 2010. Given that our data commences from December 2010, it can be a challenging task to accurately link a cancelled order with a specific previous order. Here's a simplified approach I'm using to address this issue: I start by looping from the closest previous order that matches the cancelled order. If the sum of the cancelled order's quantity and the order's cancelledQuantity is less than the order's quantity, I simply add the quantity of the cancelled order to the cancelledQuantity. However, if this isn't the case, I equalize the order's cancelledQuantity to the order's quantity, and then update the cancelled order's quantity to reflect the difference.

In [51]:
tran_before = []
tran_cur = []

df_cleaning['CancelledQuantity'] = 0

for index, col in df_cleaning[df_cleaning['InvoiceNo'].str.contains('C')].iterrows():
    previous_order = df_cleaning[(df_cleaning['CustomerID'] == col['CustomerID']) & 
                (df_cleaning['StockCode'] == col['StockCode']) & 
                (df_cleaning['InvoiceDate'] < col['InvoiceDate']) & 
                (df_cleaning['Quantity'] > 0)].copy()
    if (previous_order.shape[0] == 0):
        tran_before.append(index)
    else:
        cur_quantity = -col['Quantity']
        # tran_cur.extend(list(previous_order.index))
        for i in list(previous_order.index)[::-1]:
            if df_cleaning.loc[i, 'CancelledQuantity'] >= df_cleaning.loc[i, 'Quantity']:
                continue
            if df_cleaning.loc[i, 'Quantity'] >= df_cleaning.loc[i, 'CancelledQuantity'] + cur_quantity:
                df_cleaning.loc[i, 'CancelledQuantity'] += cur_quantity
                break
            else:
                cur_quantity -= (df_cleaning.loc[i, 'Quantity'] - df_cleaning.loc[i, 'CancelledQuantity'])
                df_cleaning.loc[i, 'CancelledQuantity'] = df_cleaning.loc[i, 'Quantity']
                
        tran_cur.append(index)

In [52]:
len(tran_before)

1303

In [53]:
tran_cur = list(set(tran_cur))
len(tran_cur)

7569

In [54]:
df_cleaning.drop(tran_before, inplace = True)
df_cleaning.drop(tran_cur, inplace = True)

# 4. Export Cleaning Data

In [55]:
df_cleaning.to_csv('../data/cleaning_data.csv', index = False)